File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
import scala.collection.mutable.Stack


object Solution {
    def maxDepth(root: TreeNode): Int = {
        val stack = new Stack[(TreeNode, Int)]()
        if (root != null) stack.push((root, 1))

        var maxDepth = 0
        while (stack.nonEmpty) {
            val (node, depth) = stack.pop()
            if (node != null) {
                maxDepth = maxDepth.max(depth)
                stack.push((node.left, depth + 1))
                stack.push((node.right, depth + 1))
            }
        }
        
        maxDepth
    }
}

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        stack = [(root, False, 1)]
        ans = 0
        while stack:
            node, visited, depth = stack.pop()
            if node and not visited:
                ans = max(ans, depth)
                stack.append((node.right, False, depth + 1))
                stack.append((node.left, False, depth + 1))
        return ans